subsys/bluetooth/host/hci_core.c

```diff
@@ -3638,9 +3638,18 @@ static void hci_vs_init(void)
 static int hci_init(void)
 {
 	int err;
+
 #if defined(CONFIG_BT_HCI_SETUP)
+	struct bt_hci_setup_params setup_params = { 0 };
+
+#if defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR)
+	if (bt_dev.id_count > 0 && bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_PUBLIC) {
+		setup_params.public_addr = &bt_dev.id_addr[BT_ID_DEFAULT].a;
+	}
+#endif /* defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR) */
+
 	if (bt_dev.drv->setup) {
-		err = bt_dev.drv->setup();
+		err = bt_dev.drv->setup(&setup_params);
```

Comment:
@arbrauns, I think this code is executed during initialization, but if the host send an HCI_RESET, this setting is lost in BlueNRG. I am not sure whether you already took into account this case (see also `bt_spi_send`, to check how HCI_RESET is managed).

Reply:
I had taken this into account initially (setting the public address in `bt_spi_handle_vendor_evt`), but it was lost with the switch to generic HCI API functions. I guess I could still save the address in `bt_spi_bluenrg_setup` and then restore it in `bt_spi_handle_vendor_evt`?

Reply:
> I had taken this into account initially (setting the public address in `bt_spi_handle_vendor_evt`), but it was lost with the switch to generic HCI API functions. I guess I could still save the address in `bt_spi_bluenrg_setup` and then restore it in `bt_spi_handle_vendor_evt`?

That sounds like a potential solution to me (though I think the question was directed at @msmttchr ).

Another option is to set `hci_driver.quirks = BT_QUIRK_NO_RESET` - that will cause the host to never send the `HCI_Reset` command.

Reply:
> I had taken this into account initially (setting the public address in `bt_spi_handle_vendor_evt`), but it was lost with the switch to generic HCI API functions. I guess I could still save the address in `bt_spi_bluenrg_setup` and then restore it in `bt_spi_handle_vendor_evt`?

Well, I thought you could use read_config prior to host reset to save public address and restore it after reset is complete, but I also seen @jhedberg comment on BT_QUIRK_NO_RESET, so I am no longer sure this extra complication is beneficial.

Reply:
> > I had taken this into account initially (setting the public address in `bt_spi_handle_vendor_evt`), but it was lost with the switch to generic HCI API functions. I guess I could still save the address in `bt_spi_bluenrg_setup` and then restore it in `bt_spi_handle_vendor_evt`?
> 
> 
> 
> That sounds like a potential solution to me (though I think the question was directed at @msmttchr ).
> 
> 
> 
> Another option is to set `hci_driver.quirks = BT_QUIRK_NO_RESET` - that will cause the host to never send the `HCI_Reset` command.

Ah ah, I think this is a better proposal to keep things reasonably linear. I initially thought that this setting prevented reset only during initialization, but if this is meant to prevent sending HCI reset in all cases, I think this is just fine.